### PR TITLE
Push metrics

### LINKED ISSUE #747: make sure that yagna have terms and conditions containing opt-in for monitoring

### PR DESCRIPTION
**Pull request by jiivan**
fixes: #720
fixes #747

### CODE REVIEW

File: core/metrics/src/service.rs [hunk 1]
```diff
@@ -38,3 +40,25 @@ impl MetricsService {
 pub async fn expose_metrics(metrics: web::Data<Arc<Mutex<Metrics>>>) -> impl Responder {
     metrics.lock().await.export()
 }
+
+const PUSH_HOST_URL: &str = "http://127.0.0.1:9091/";
```
Comment (nieznanysprawiciel) [inline core/metrics/src/service.rs:44]: Better to have it configurable

**Review by nieznanysprawiciel — APPROVED** [review]

File: core/metrics/src/service.rs [hunk 2]
```diff
@@ -38,3 +40,25 @@ impl MetricsService {
...
+        .call(identity::Get::ByDefault)
+        .await??
+        .ok_or(anyhow::anyhow!("Default identity not found"))?;
+    let base = url::Url::parse(host_url)?;
+    let url = base.join("/metrics/job/community.1/instance/")?.join(format!("{}", default_id.node_id).as_str())?;
+    Ok(String::from(url.as_str()))
+}
+
+pub async fn push() {
+    let push_url = get_push_url(PUSH_HOST_URL).await.unwrap();
```
Comment (tworec) [inline core/metrics/src/service.rs:57]: make it log::debug and skip pushing

File: core/metrics/src/metrics.rs [hunk 3]
```diff
@@ -35,3 +43,13 @@ impl Metrics {
         return self.exporter.turn();
     }
 }
+
+async fn push_forever() {
+    let mut interval = time::interval(Duration::from_secs(5));
+    loop {
+        interval.tick().await;
+        log::error!("Metrics push job started");
```
Comment (tworec) [inline core/metrics/src/metrics.rs:51]:
```suggestion
```

File: core/metrics/src/metrics.rs [hunk 4]
```diff
@@ -35,3 +43,13 @@ impl Metrics {
...
+        crate::service::push().await;
+        log::debug!("Metrics push job done");
```
Comment (tworec) [inline core/metrics/src/metrics.rs:53]:
```suggestion
```

File: core/metrics/src/metrics.rs [hunk 5]
```diff
@@ -19,6 +21,12 @@ impl Metrics {
         let exporter = StringExporter::new(receiver.controller(), PrometheusBuilder::new());
         receiver.install();
 
+        log::error!("Starting pusher");
+        tokio::task::spawn_local(async move {
+            push_forever().await;
+        });
+        log::error!("Started pusher");
```
Comment (tworec) [inline core/metrics/src/metrics.rs:28]:
```suggestion
        log::info!("Started Metric pusher");
```

File: core/metrics/src/service.rs [hunk 6]
```diff
@@ -38,3 +40,25 @@ impl MetricsService {
...
+    let current_metrics = METRICS.clone().lock().await.export();
+    log::error!("Got current metrics");
```
Comment (tworec) [inline core/metrics/src/service.rs:59]:
```suggestion
```

File: core/metrics/src/service.rs [hunk 7]
```diff
@@ -38,3 +40,25 @@ impl MetricsService {
...
+    let client = reqwest::Client::new();
+    let res = client.put(push_url.as_str())
+        .body(current_metrics).send().await;
+    log::error!("Pushed current metrics {:#?}", res);
```
Comment (tworec) [inline core/metrics/src/service.rs:63]:
```suggestion
    log::trace!("Pushed current metrics {:#?}", res);
```

**Review by tworec — CHANGES_REQUESTED** [review]

File: core/metrics/src/service.rs (hunk 1 above)
Comment (tworec) [reply core/metrics/src/service.rs:44]: env var + defgault to `http://metrics.golem.network:9091/`

File: core/metrics/src/service.rs [hunk 8]
```diff
@@ -38,3 +40,25 @@ impl MetricsService {
 pub async fn expose_metrics(metrics: web::Data<Arc<Mutex<Metrics>>>) -> impl Responder {
     metrics.lock().await.export()
 }
+
+const PUSH_HOST_URL: &str = "http://127.0.0.1:9091/";
+
+async fn get_push_url(host_url: &str) -> anyhow::Result<String> {
```
Comment (tworec) [inline core/metrics/src/service.rs:46]: do it once, before loop

**Comment by tworec** [comment]
add an configuration flag to disable metrics sending (we are sending them by default)

File: core/serv-api/src/lib.rs [hunk 9]
```diff
@@ -3,13 +3,20 @@ use prettytable::{color, format, format::TableFormat, Attr, Cell, Row, Table};
 use serde::Serialize;
 use std::path::PathBuf;
 
+#[derive(Debug, Default)]
+pub struct MetricsCtx {
+    pub push_enabled: bool,
+    pub push_host_url: Option<url::Url>,
```
[the remark this answers is not in the document]
Comment (jiivan) [reply core/serv-api/src/lib.rs:9]: It's needed because of `Default`

**Review by tworec — DISMISSED** [review]

File: core/metrics/src/pusher.rs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+async fn get_push_url(host_url: &str, id: IdentityInfo) -> anyhow::Result<String> {
+    let base = url::Url::parse(host_url)?;
+    let url = base
+        .join("/metrics/job/community.1/")?
+        .join(&format!("instance/{}/", id.node_id))?
+        .join(&format!(
+            "hostname/{}",
+            id.alias.unwrap_or(id.node_id.to_string())
```
Comment (jiivan) [inline core/metrics/src/pusher.rs:81]: I'm not sure about using user generated names here. We should at least [encode them](https://docs.rs/percent-encoding/2.1.0/percent_encoding/)

Comment (tworec) [reply core/metrics/src/pusher.rs:81]: good catch! thx. will fix

**Review by tworec — APPROVED** [review]

File: core/serv-api/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tworec) [reply core/serv-api/src/lib.rs:9]: Ah, I see it is for supporting Default derived impl

**Pull request merged**